### PR TITLE
Loan protocol

### PR DESCRIPTION
**Pull request by luckysori**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -216,3 +216,19 @@ jobs:
...
+
+      - name: Install Rust toolchain
+        id: toolchain
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          override: true
+
+      - name: Cargo test
+        run: cargo test -p covenants
```
Comment (luckysori) [inline .github/workflows/ci.yml:234]: Ooops, we don't actually need this. There already was a job executing cargo tests for the whole workspace.

Comment (bonomat) [reply .github/workflows/ci.yml:234]: True that :)

File: covenants/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,1160 @@
...
+                        && out.script_pubkey == self.address.script_pubkey();
+
+                    predicate.then(|| out)
+                }
+                None => None,
+            })
+            .context("no collateral change txout")?;
+
+        Ok(Borrower1 {
+            keypair: self.keypair,
```
Comment (bonomat) [inline covenants/src/lib.rs:211]: nit: renaming to what this key is used.

File: covenants/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,1160 @@
...
+            tx.input[1].witness = TxInWitness {
+                amount_rangeproof: vec![],
+                inflation_keys_rangeproof: vec![],
+                script_witness,
+                pegin_witness: vec![],
+            };
+        };
+
+        // TODO: Sign more than one input if necessary
+        // sign repayment input of the principal amount
```
Comment (bonomat) [inline covenants/src/lib.rs:444]: No need to do this now, but once extracted into a lib we should do this!

File: covenants/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,1160 @@
...
+impl Lender0 {
+    pub fn new<C>(
+        secp: &Secp256k1<C>,
+        bitcoin_asset_id: AssetId,
+        usdt_asset_id: AssetId,
+        // TODO: Here we assume that the wallet is giving us _all_ the
+        // inputs available. It would be better to coin-select these
+        // as soon as we know the principal amount after receiving the
+        // loan request
+        principal_inputs: Vec<Input>,
```
Comment (bonomat) [inline covenants/src/lib.rs:477]: We should pass in a coin-selector as in the other function e.g. see `borrower.loan_repayment_transaction(..)`.

File: covenants/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,1160 @@
...
+                collateral_change_tx_out,
+                tx_fee_tx_out,
+            ],
+        };
+
+        let repayment_collateral_input = Input {
+            tx_in: TxIn {
+                previous_output: OutPoint {
+                    txid: loan_transaction.txid(),
+                    vout: 0,
```
Comment (bonomat) [inline covenants/src/lib.rs:684]: This looks a bit dangerous to me. 😬

File: covenants/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,1160 @@
...
+    lender_pk: PublicKey,
+    repayment_collateral_input: Input,
+    repayment_collateral_abf: AssetBlindingFactor,
+    repayment_collateral_vbf: ValueBlindingFactor,
+    timelock: u64,
+    repayment_principal_output: TxOut,
+}
+
+pub struct Borrower0 {
+    keypair: (SecretKey, PublicKey),
```
Comment (bonomat) [inline covenants/src/lib.rs:70]: nit: do we need to keep the `PublicKey` around?

File: covenants/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,1160 @@
...
+        // input specific values
+        let (previous_out, script_0, script_1, script_2, value, sequence) = {
+            let InputData {
+                previous_output,
+                script,
+                value,
+                sequence,
+            } = &self.input;
+
+            let third = script.len() / 3;
```
Comment (bonomat) [inline covenants/src/lib.rs:1007]: We should fail here if `script.len()/3 > 80`

File: covenants/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,1160 @@
...
+
+            if self.other_outputs.len() > 3 {
+                bail!("too many outputs");
+            }
+
+            for txout in self.other_outputs.iter() {
+                let mut output = Vec::new();
+                txout.consensus_encode(&mut output)?;
+
+                let middle = output.len() / 2;
```
Comment (bonomat) [inline covenants/src/lib.rs:1059]: Same as above, fail if `output.len()/2>80`

File: covenants/src/protocol_tests.rs [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+    use elements_harness::{elementd_rpc::ElementsRpc, Elementsd};
+    use secp256k1_zkp::Message;
+    use std::env;
+    use testcontainers::clients::Cli;
+
+    #[tokio::test]
+    async fn borrow_and_repay() {
+        // force enabling log output
+        // TODO: remove me again
+        env::set_var("RUST_LOG", "DEBUG");
```
Comment (bonomat) [inline covenants/src/protocol_tests.rs:24]: Can be removed now.

File: covenants/src/protocol_tests.rs [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+                timelock,
+                bitcoin_asset_id,
+                usdt_asset_id,
+            )
+            .unwrap();
+
+            (borrower, address)
+        };
+
+        // TODO: Use a separate wallet per actor. Using the same wallet is confusing and bug-prone.
```
Comment (bonomat) [inline covenants/src/protocol_tests.rs:86]: What do you think @luckysori : shall we tackle this before merging? 
Could be easier to find potential flaws now than later.

File: covenants/src/protocol_tests.rs [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+        let (lender, _lender_address) = {
+            let address = client
+                .get_new_address(Some("blech32".into()))
```
Comment (bonomat) [inline covenants/src/protocol_tests.rs:89]: we could add a helper function to return a confidential address which is in blech32 format, e.g. 
`client.get_new_confidential_address()`.

File: covenants/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,1160 @@
...
+                .encode_segwitv0_signing_data_to(
+                    &mut unhashed,
+                    1,
+                    &self.collateral_script.clone(),
+                    self.repayment_collateral_input.original_tx_out.value,
+                    SigHashType::All,
+                )
+                .unwrap();
+
+            dbg!(hex::encode(unhashed));
```
Comment (bonomat) [inline covenants/src/lib.rs:406]: Can be removed now.

File: covenants/src/protocol_tests.rs [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+            .await
+            .unwrap();
+
+        client
+            .send_raw_transaction(&loan_repayment_transaction)
+            .await
+            .unwrap();
+    }
+
+    // #[tokio::test]
```
Comment (bonomat) [inline covenants/src/protocol_tests.rs:194]: Ups, completely forgot about this :D

File: covenants/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,1142 @@
...
+
+        let sighash_type = {
+            let mut writer = Vec::new();
+            self.sighash_type.as_u32().consensus_encode(&mut writer)?;
+            writer
+        };
+
+        Ok(vec![
+            sig,
+            pk,
```
Comment (luckysori) [inline covenants/src/lib.rs:1069]: @bonomat remind me: why do we need this pk in the witness stack? Don't we already have it in the script? Is it just for convenience?

File: covenants/src/lib.rs (hunk 2 above)
Comment (luckysori) [reply covenants/src/lib.rs:209]: What did you have in mind? Isn't this just the borrower's keypair?

File: covenants/src/lib.rs (hunk 6 above)
Comment (luckysori) [reply covenants/src/lib.rs:70]: We use it both the lender's and borrower's PKs every time we compute the loan contract, so I think we do.

File: covenants/src/protocol_tests.rs [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+
+        client
+            .send_raw_transaction(&loan_repayment_transaction)
+            .await
+            .unwrap();
+    }
+
+    fn init_logger() {
+        // force enabling log output
+        // TODO: remove me again
```
Comment (luckysori) [inline covenants/src/protocol_tests.rs:193]: :question: @bonomat So do you want to remove this helper function or keep it?

File: covenants/src/lib.rs (hunk 2 above)
Comment (bonomat) [reply covenants/src/lib.rs:185]: Yes, but for what? I honestly forgot for what atm :D

File: covenants/src/protocol_tests.rs (hunk 15 above)
Comment (bonomat) [reply covenants/src/protocol_tests.rs:330]: Keep the `init_logger` function but remove the next line: `env::set_var("RUST_LOG", "DEBUG");` which should be set from the outside

File: covenants/src/lib.rs (hunk 14 above)
Comment (bonomat) [reply covenants/src/lib.rs:1049]: You are right, we have it already in the script `        .push_slice(&borrower_pk.serialize())
`. 
I don't think there was a particular reason for having it here again. IF we change it, we will need to make use of `OP_DUP` and move the `pk` in the script around a couple of times or add it to the script twice. (we need it for `op_checksig` and `op_checksigfromstack`).

Comment (luckysori) [reply covenants/src/lib.rs:1047]: See #155

**Pull request merged**